File: src/main/java/com/myudog/myulib/api/VFXCompat.java
package com.myudog.myulib.api;
import net.minecraft.core.particles.ParticleOptions;
import net.minecraft.server.level.ServerLevel;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
public final class VFXCompat {
    private VFXCompat() {
    }
    public static void spawnParticles(ServerLevel Level, ParticleOptions effect, double x, double y, double z, int count, double offsetX, double offsetY, double offsetZ, double speed) {
        Method[] methods = ServerLevel.class.getDeclaredMethods();
        for (Method method : methods) {
            if (!method.getName().equals("spawnParticles")) {
                continue;
            }
            Class<?>[] types = method.getParameterTypes();
            if (types.length == 9 && matches(types, 0, ParticleOptions.class) && matches(types, 1, double.class) && matches(types, 2, double.class) && matches(types, 3, double.class)) {
                try {
                    method.setAccessible(true);
                    if (matches(types, 4, int.class)) {
                        method.invoke(Level, effect, x, y, z, count, offsetX, offsetY, offsetZ, speed);
                        return;
                    }
                    if (matches(types, 8, int.class)) {
                        method.invoke(Level, effect, x, y, z, offsetX, offsetY, offsetZ, speed, count);
                        return;
                    }
                } catch (Throwable ignored) {
                    // try next candidate
                }
            }
            if (types.length == 8 && matches(types, 0, ParticleOptions.class)) {
                try {
                    method.setAccessible(true);
                    for (int i = 0; i < count; i++) {
                        method.invoke(Level, effect, x, y, z, offsetX, offsetY, offsetZ, speed);
                    }
                    return;
                } catch (Throwable ignored) {
                    // try next candidate
                }
            }
        }
        throw new NoSuchMethodError("No suitable spawnParticles overload found for this environment");
    }
    private static boolean matches(Class<?>[] types, int index, Class<?> expected) {
        return index >= 0 && index < types.length && types[index].equals(expected);
    }
}

File: src/main/kotlin/com/myudog/myulib/api/VFXCompat.kt
package com.myudog.myulib.api

import net.minecraft.particle.ParticleEffect
import net.minecraft.server.world.ServerWorld
import java.lang.reflect.Method

object VFXCompat {
    private val spawnMethods: List<Method> by lazy {
        val methods = mutableListOf<Method>()
        ServerWorld::class.java.declaredMethods.filter { it.name == "spawnParticles" }.forEach { methods += it }
        methods
    }

    /**
     * Try to invoke ServerWorld.spawnParticles with best-matching signature.
     * We accept the common parameters and dispatch reflectively to handle mapping differences.
     */
    fun spawnParticles(
        world: ServerWorld,
        effect: ParticleEffect,
        x: Double,
        y: Double,
        z: Double,
        count: Int,
        offsetX: Double,
        offsetY: Double,
        offsetZ: Double,
        speed: Double
    ) {
        // Candidate signatures we support (param type arrays)
        val candidates = listOf(
            arrayOf(ParticleEffect::class.java, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType, Int::class.javaPrimitiveType, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType),
            arrayOf(ParticleEffect::class.java, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType, Double::class.javaPrimitiveType, Int::class.javaPrimitiveType)
        )

        for (m in spawnMethods) {
            val ptypes = m.parameterTypes
            for (cand in candidates) {
                if (ptypes.size == cand.size && ptypes.indices.all { ptypes[it] == cand[it] }) {
                    try {
                        m.isAccessible = true
                        if (cand[4] == Int::class.javaPrimitiveType) {
                            m.invoke(world, effect, x, y, z, count, offsetX, offsetY, offsetZ, speed)
                        } else {
                            // count is last
                            m.invoke(world, effect, x, y, z, offsetX, offsetY, offsetZ, speed, count)
                        }
                        return
                    } catch (_: Throwable) {
                        // try next
                    }
                }
            }
        }

        // If none matched, try to find a spawnParticles with 8 params (without count) and call multiple times
        val eight = spawnMethods.firstOrNull { it.parameterCount == 8 }
        if (eight != null) {
            try {
                eight.isAccessible = true
                // (effect,x,y,z,dx,dy,dz,speed) -> call with dx,dy,dz,speed and count times
                for (i in 0 until count) {
                    eight.invoke(world, effect, x, y, z, offsetX, offsetY, offsetZ, speed)
                }
                return
            } catch (_: Throwable) {}
        }

        // As last resort, throw
        throw NoSuchMethodError("No suitable spawnParticles overload found for this environment")
    }
}

